lib/src/wallet/traits.rs

```diff
@@ -0,0 +1,590 @@
+        keys.get_all_sapling_addresses()
+    }
+
+    fn get_keys(keys: &Keys) -> &Vec<Self> {
+        keys.zkeys()
+    }
+
+    fn set_spend_key_for_view_key(&mut self, key: Self::Sk) {
+        self.extsk = Some(key);
+        self.keytype = super::keys::sapling::WalletZKeyType::ImportedSpendingKey;
+    }
+}
+
+impl WalletKey for OrchardKey {
+    type Sk = OrchardSpendingKey;
+
+    type Fvk = OrchardFullViewingKey;
+
+    type Ivk = OrchardIncomingViewingKey;
+
+    type Ovk = OrchardOutgoingViewingKey;
+
+    type Address = zcash_address::unified::Address;
+
+    fn sk(&self) -> Option<Self::Sk> {
+        (&self.key).try_into().ok()
+    }
+    fn fvk(&self) -> Option<Self::Fvk> {
+        (&self.key).try_into().ok()
+    }
+    fn ivk(&self) -> Option<Self::Ivk> {
+        (&self.key).try_into().ok()
+    }
+
+    fn ovk(&self) -> Option<Self::Ovk> {
+        (&self.key).try_into().ok()
+    }
+
+    fn address(&self) -> Self::Address {
+        self.unified_address.clone()
+    }
+
+    fn addresses_from_keys(keys: &Keys) -> Vec<String> {
+        keys.get_all_orchard_addresses()
+    }
+
+    fn get_keys(keys: &Keys) -> &Vec<Self> {
+        keys.okeys()
+    }
+
+    fn set_spend_key_for_view_key(&mut self, key: Self::Sk) {
+        self.key = super::keys::orchard::WalletOKeyInner::ImportedSpendingKey(key)
+    }
+}
+
+impl NoteData for OrchardNoteData {
+    type Fvk = OrchardFullViewingKey;
+    type Diversifier = OrchardDiversifier;
+    type Note = OrchardNote;
+    type Node = MerkleHashOrchard;
+    type Nullifier = OrchardNullifier;
+
+    fn from_parts(
+        fvk: Self::Fvk,
+        diversifier: Self::Diversifier,
+        note: Self::Note,
+        witnesses: WitnessCache<Self::Node>,
+        nullifier: Self::Nullifier,
+        spent: Option<(TxId, u32)>,
+        unconfirmed_spent: Option<(TxId, u32)>,
+        memo: Option<Memo>,
+        is_change: bool,
+        have_spending_key: bool,
+    ) -> Self {
+        Self {
+            fvk,
+            diversifier,
+            note,
+            witnesses,
+            nullifier,
+            spent,
+            unconfirmed_spent,
+            memo,
+            is_change,
+            have_spending_key,
+        }
+    }
+
+    fn memo_mut(&mut self) -> &mut Option<Memo> {
+        &mut self.memo
+    }
+    fn note(&self) -> &Self::Note {
+        &self.note
+    }
+    fn nullifier(&self) -> Self::Nullifier {
+        self.nullifier
+    }
+
+    fn value(note: &Self::Note) -> u64 {
+        note.value().inner()
+    }
+
+    fn wallet_transaction_notes_mut(wallet_transaction: &mut WalletTx) -> &mut Vec<Self> {
+        &mut wallet_transaction.orchard_notes
+    }
+
+    fn witnesses(&mut self) -> &mut WitnessCache<Self::Node> {
+        &mut self.witnesses
+    }
+}
+
+pub(crate) trait DomainWalletExt<P: Parameters>: Domain
+where
+    Self: Sized,
+    Self::Note: PartialEq,
+    Self::Recipient: Recipient,
+    Self::Note: PartialEq,
+{
+    type Fvk: Clone;
```

Comment:
Given that `zcash_note_encryption::Domain` has IVK and OVK associated types..  is it really necessary to also have the FVK?